When building `-A nix` via nix 2.3 on 21b3094ca2de8906aa19f83c855959988e9b12e4, the error looks like this:

```
ran test tests/ca/substitute.sh... [FAIL]
    ++++(/build/source/tests/common/vars-and-functions.sh:274) trap onError ERR
    +++(../common.sh:8) [[ -n '' ]]
    ++(common.sh:3) enableFeatures ca-derivations
    ++(/build/source/tests/common/vars-and-functions.sh:230) local features=ca-derivations
    ++(/build/source/tests/common/vars-and-functions.sh:231) sed -i 's/experimental-features .*/& ca-derivations/' /build/nix-test/tests/ca/substitute/etc/nix.conf
    ++(common.sh:5) restartDaemon
    ++(/build/source/tests/common/vars-and-functions.sh:138) [[ -z '' ]]
    ++(/build/source/tests/common/vars-and-functions.sh:138) return 0
    +(substitute.sh:7) needLocalStore '“--no-require-sigs” can’t be used with the daemon'
    +(/build/source/tests/common/vars-and-functions.sh:219) [[ '' == \d\a\e\m\o\n ]]
    +(substitute.sh:9) rm -rf /build/nix-test/tests/ca/substitute/binary_cache
    +(substitute.sh:11) export REMOTE_STORE_DIR=/build/nix-test/tests/ca/substitute/binary_cache
    +(substitute.sh:11) REMOTE_STORE_DIR=/build/nix-test/tests/ca/substitute/binary_cache
    +(substitute.sh:12) export REMOTE_STORE=file:///build/nix-test/tests/ca/substitute/binary_cache
    +(substitute.sh:12) REMOTE_STORE=file:///build/nix-test/tests/ca/substitute/binary_cache
    +(substitute.sh:19) clearStore
    +(/build/source/tests/common/vars-and-functions.sh:74) echo 'clearing store...'
    clearing store...
    +(/build/source/tests/common/vars-and-functions.sh:75) chmod -R +w /build/nix-test/tests/ca/substitute/store
    +(/build/source/tests/common/vars-and-functions.sh:76) rm -rf /build/nix-test/tests/ca/substitute/store
    +(/build/source/tests/common/vars-and-functions.sh:77) mkdir /build/nix-test/tests/ca/substitute/store
    +(/build/source/tests/common/vars-and-functions.sh:78) rm -rf /build/nix-test/tests/ca/substitute/var/nix
    +(/build/source/tests/common/vars-and-functions.sh:79) mkdir /build/nix-test/tests/ca/substitute/var/nix
    +(/build/source/tests/common/vars-and-functions.sh:80) clearProfiles
    +(/build/source/tests/common/vars-and-functions.sh:69) profiles=/build/nix-test/tests/ca/substitute/test-home/.local/state/nix/profiles
    +(/build/source/tests/common/vars-and-functions.sh:70) rm -rf /build/nix-test/tests/ca/substitute/test-home/.local/state/nix/profiles
    +(substitute.sh:20) nix copy --to file:///build/nix-test/tests/ca/substitute/binary_cache --file ./content-addressed.nix
    warning: you don't have Internet access; disabling some network-dependent features
    these 7 derivations will be built:
      /build/nix-test/tests/ca/substitute/store/5ij4pc9x6hafh35zb0lfvdfr142dqd1k-simple-input-addressed.drv
      /build/nix-test/tests/ca/substitute/store/9hdnchxn0qd08671ld2dq3nxdb9jvf80-rootCA.drv
      /build/nix-test/tests/ca/substitute/store/xlyjw5k36dkz4a06qsbg8qvvwcasw7yv-dependent.drv
      /build/nix-test/tests/ca/substitute/store/b1n5cknf76m68b3wk5lkycayf1a0z5lj-dependent-fixed-output.drv
      /build/nix-test/tests/ca/substitute/store/cl3j77wq9r3mazp0h6fr6fb20c491kly-transitively-dependent.drv
      /build/nix-test/tests/ca/substitute/store/v1y33skffczjalqdl37qp5gmwrmh0jcb-dependent-non-ca.drv
      /build/nix-test/tests/ca/substitute/store/zzj2q12i6mhcla645g7qgrn50ypf17vs-runnable-thing.drv
    building '/build/nix-test/tests/ca/substitute/store/5ij4pc9x6hafh35zb0lfvdfr142dqd1k-simple-input-addressed.drv'...
    building '/build/nix-test/tests/ca/substitute/store/272sjfm68kpv7wvcfl3k5c676rdgwhwm-rootCA.drv'...
    building '/build/nix-test/tests/ca/substitute/store/ippsy51k6y49gzml56xz8arj4cgbw1w1-dependent-non-ca.drv'...
    building '/build/nix-test/tests/ca/substitute/store/50c5ssa92978jpn2d4ivrybq5xs8g2h2-dependent.drv'...
    building '/build/nix-test/tests/ca/substitute/store/2s5jwgpl9ay20bva6kqhigwsr5crhvch-runnable-thing.drv'...
    building '/build/nix-test/tests/ca/substitute/store/mx4izwcknc113ki1cvh903s0f1p9zhi3-dependent-fixed-output.drv'...
    building '/build/nix-test/tests/ca/substitute/store/6xdz6jfp3b2p54dh9sn1kyn7j29djqcq-transitively-dependent.drv'...
    copying 9 paths...
    copying path '/build/nix-test/tests/ca/substitute/store/3vxm659ffkp26lij9s611nxh490zdwsp-dependent-fixed-output' to 'file:///build/nix-test/tests/ca/substitute/binary_cache'...
    copying path '/build/nix-test/tests/ca/substitute/store/xf6l51diijwn41biq6smhvvgzxw47k7x-simple-input-addressed' to 'file:///build/nix-test/tests/ca/substitute/binary_cache'...
    copying path '/build/nix-test/tests/ca/substitute/store/9mmvwhxf5hzc4xk04vyzgmk2qvsgd57p-runnable-thing' to 'file:///build/nix-test/tests/ca/substitute/binary_cache'...
    copying path '/build/nix-test/tests/ca/substitute/store/p47vs5ka8w3sildrynvz2myyqlkiymdy-rootCA' to 'file:///build/nix-test/tests/ca/substitute/binary_cache'...
    copying path '/build/nix-test/tests/ca/substitute/store/14f4vgcwyibbz9qnckp6g6bfbbd3y10p-rootCA-foo' to 'file:///build/nix-test/tests/ca/substitute/binary_cache'...
    copying path '/build/nix-test/tests/ca/substitute/store/ypkzca05ab09mj7jlddjvsj6zbmxzm3r-rootCA-dev' to 'file:///build/nix-test/tests/ca/substitute/binary_cache'...
    copying path '/build/nix-test/tests/ca/substitute/store/10dx3bxpb716ajq54rbd30dzs629h2b5-dependent-non-ca' to 'file:///build/nix-test/tests/ca/substitute/binary_cache'...
    copying path '/build/nix-test/tests/ca/substitute/store/vqypk04lb0yxf1j4yx5d7z6zf954mmdi-dependent' to 'file:///build/nix-test/tests/ca/substitute/binary_cache'...
    copying path '/build/nix-test/tests/ca/substitute/store/iqrvx8i366z4iyh5wfvwjvagn7fg0djj-transitively-dependent' to 'file:///build/nix-test/tests/ca/substitute/binary_cache'...
    +(substitute.sh:23) clearStore
    +(/build/source/tests/common/vars-and-functions.sh:74) echo 'clearing store...'
    clearing store...
    +(/build/source/tests/common/vars-and-functions.sh:75) chmod -R +w /build/nix-test/tests/ca/substitute/store
    +(/build/source/tests/common/vars-and-functions.sh:76) rm -rf /build/nix-test/tests/ca/substitute/store
    +(/build/source/tests/common/vars-and-functions.sh:77) mkdir /build/nix-test/tests/ca/substitute/store
    +(/build/source/tests/common/vars-and-functions.sh:78) rm -rf /build/nix-test/tests/ca/substitute/var/nix
    +(/build/source/tests/common/vars-and-functions.sh:79) mkdir /build/nix-test/tests/ca/substitute/var/nix
    +(/build/source/tests/common/vars-and-functions.sh:80) clearProfiles
    +(/build/source/tests/common/vars-and-functions.sh:69) profiles=/build/nix-test/tests/ca/substitute/test-home/.local/state/nix/profiles
    +(/build/source/tests/common/vars-and-functions.sh:70) rm -rf /build/nix-test/tests/ca/substitute/test-home/.local/state/nix/profiles
    +(substitute.sh:24) buildDrvs --substitute --substituters file:///build/nix-test/tests/ca/substitute/binary_cache --no-require-sigs -j0 transitivelyDependentCA
    +(substitute.sh:15) nix build --file ./content-addressed.nix -L --no-link --substitute --substituters file:///build/nix-test/tests/ca/substitute/binary_cache --no-require-sigs -j0 transitivelyDependentCA
    warning: you don't have Internet access; disabling some network-dependent features
    these 3 derivations will be built:
      /build/nix-test/tests/ca/substitute/store/9hdnchxn0qd08671ld2dq3nxdb9jvf80-rootCA.drv
      /build/nix-test/tests/ca/substitute/store/xlyjw5k36dkz4a06qsbg8qvvwcasw7yv-dependent.drv
      /build/nix-test/tests/ca/substitute/store/cl3j77wq9r3mazp0h6fr6fb20c491kly-transitively-dependent.drv
    this path will be fetched (0.00 MiB download, 0.00 MiB unpacked):
      /build/nix-test/tests/ca/substitute/store/xf6l51diijwn41biq6smhvvgzxw47k7x-simple-input-addressed
    copying path '/build/nix-test/tests/ca/substitute/store/xf6l51diijwn41biq6smhvvgzxw47k7x-simple-input-addressed' from 'file:///build/nix-test/tests/ca/substitute/binary_cache'...
    copying path '/build/nix-test/tests/ca/substitute/store/p47vs5ka8w3sildrynvz2myyqlkiymdy-rootCA' from 'file:///build/nix-test/tests/ca/substitute/binary_cache'...
    copying path '/build/nix-test/tests/ca/substitute/store/vqypk04lb0yxf1j4yx5d7z6zf954mmdi-dependent' from 'file:///build/nix-test/tests/ca/substitute/binary_cache'...
    copying path '/build/nix-test/tests/ca/substitute/store/iqrvx8i366z4iyh5wfvwjvagn7fg0djj-transitively-dependent' from 'file:///build/nix-test/tests/ca/substitute/binary_cache'...
    +(substitute.sh:26) nix realisation info --file ./content-addressed.nix transitivelyDependentCA
    warning: you don't have Internet access; disabling some network-dependent features
    sha256:3e4163dea7e3bac8077836091aabe1c81dd14b8ba8d6690dcc8ccbaef5e94cd0!out /build/nix-test/tests/ca/substitute/store/iqrvx8i366z4iyh5wfvwjvagn7fg0djj-transitively-dependent
    +(substitute.sh:28) nix realisation info --file ./content-addressed.nix dependentCA
    warning: you don't have Internet access; disabling some network-dependent features
    sha256:880c5679eee6c9631c8010d4d108d6bab79366721ebc4403d044389ee49aea57!out /build/nix-test/tests/ca/substitute/store/vqypk04lb0yxf1j4yx5d7z6zf954mmdi-dependent
    +(substitute.sh:31) isDaemonNewer 2.13
    +(/build/source/tests/common/vars-and-functions.sh:149) [[ -n '' ]]
    +(/build/source/tests/common/vars-and-functions.sh:149) return 0
    +(substitute.sh:32) pushToStore=../push-to-store.sh
    +(substitute.sh:45) clearStore
    +(/build/source/tests/common/vars-and-functions.sh:74) echo 'clearing store...'
    clearing store...
    +(/build/source/tests/common/vars-and-functions.sh:75) chmod -R +w /build/nix-test/tests/ca/substitute/store
    +(/build/source/tests/common/vars-and-functions.sh:76) rm -rf /build/nix-test/tests/ca/substitute/store
    +(/build/source/tests/common/vars-and-functions.sh:77) mkdir /build/nix-test/tests/ca/substitute/store
    +(/build/source/tests/common/vars-and-functions.sh:78) rm -rf /build/nix-test/tests/ca/substitute/var/nix
    +(/build/source/tests/common/vars-and-functions.sh:79) mkdir /build/nix-test/tests/ca/substitute/var/nix
    +(/build/source/tests/common/vars-and-functions.sh:80) clearProfiles
    +(/build/source/tests/common/vars-and-functions.sh:69) profiles=/build/nix-test/tests/ca/substitute/test-home/.local/state/nix/profiles
    +(/build/source/tests/common/vars-and-functions.sh:70) rm -rf /build/nix-test/tests/ca/substitute/test-home/.local/state/nix/profiles
    +(substitute.sh:46) nix build --file ../simple.nix -L --no-link --post-build-hook ../push-to-store.sh
    warning: you don't have Internet access; disabling some network-dependent features
    this derivation will be built:
      /build/nix-test/tests/ca/substitute/store/2vbx9wjgss5a7nhrwaqqvwsj7m85z759-simple.drv
    building '/build/nix-test/tests/ca/substitute/store/2vbx9wjgss5a7nhrwaqqvwsj7m85z759-simple.drv'...
    simple> PATH=
    running post-build-hook '../push-to-store.sh'...
    simple (post)> ../push-to-store.sh: line 5: syntax error: bad substitution
    error: program '../push-to-store.sh' failed with exit code 2
    ++(substitute.sh:46) onError
    ++(/build/source/tests/common/vars-and-functions.sh:237) set +x
    substitute.sh: test failed at:
      main in substitute.sh:46
make: *** [mk/lib.mk:123: tests/ca/substitute.sh.test] Error 1
```

I suspect it is an impurity of some kind, because adding `--keep-failed`, going into the failed dir, sourcing environment, adding `export TMPDIR=$(pwd)`, and running `make installcheck` by hand works. The script in question refers to `/bin/sh`, so I guess that it either does not exist is the sandbox on 2.3, or it is outdated/broken in some way there.

--- a/pkgs/tools/package-management/nix/common.nix
+++ b/pkgs/tools/package-management/nix/common.nix
@@ -132,6 +132,10 @@ self = stdenv.mkDerivation {
     (lib.optionalString (stdenv.hostPlatform.system == "armv5tel-linux" || stdenv.hostPlatform.system == "armv6l-linux") "-latomic")
   ];
 
+  postPatch = ''
+    patchShebangs --build tests
+  '';
+
   preConfigure =
     # Copy libboost_context so we don't get all of Boost in our closure.
     # https://github.com/NixOS/nixpkgs/issues/45462

